evaluation/FEEDBACK.md

```diff
@@ -0,0 +1,61 @@
+# Feedback
+
+This document is aiming to aggregate a set of generic rules to make writing good feedback a little easier. Note that it is by far not covering everything.
+
+P.S. Don't forget that common sense goes above all guidelines. Be smart.
+
+## Purpose
+
+Each point in feedback should be aimed to helping receiving person improve. Think about it when writing each of the points.
+
+## Actionable Points
+
+Feedback loses a lot of value if it's not clear what action receiver should take. In most cases it manifests as one of two properties:
+
+A. Reference to a topic or source that covers the point.
+
+B. Suggesting an alternative.
+
+```
+Yes:
+- Design of X is not testable. One way would be to extract logic A and test it separately.
+
+No:
+- Design of X is not testable.
+```
+
+## Staying Specific
+
+To avoid ambiguity and confusion, stay as specific as you can. If having hard time rephrasing, you could add an example.
+
+```
+Yes:
+- Prone to race conditions. I.e. same ID could be picked for different records in createNewUser method.
+
+No:
+- Prone to race conditions.
+```
+
+## Staying Factual
+
+Keeping feedback factual makes it easier to consume and act on.
+
+```
+Yes:
+- Did not solve the problem without help.
+
+No:
+- Had hard time solving the problem.
+```
+
+## Targeting Criticism
+
+Criticism (when possible) should be targeted at the submission/solution and not the candidate. It helps avoiding mental blocking.
+
+```
+Yes:
+- Solution doesn't handle invalid inputs.
+
+No:
+- Didn't implement handling invalid inputs.
+```
```

Comment:
I think that sometimes, the candidates think they fulfilled some requirement we gave them, but what they have done is not sufficient/wrong. In cases when candidate does not understand the requirement itself, just stating that it was not fulfilled is not enough and creates confusion. To avoid this, I would suggest adding some clarification about what the requirement is and why it is important.

For example, the current homework assignment has `Has a test suite. We will pay attention to coverage, structuring and flexibility.` as a requirement. It is pretty vague, says nothing about having unit tests, integration tests although we expect the solution to have both. This leaves some room for misinterpreting the requirements.

```
Yes:
- The code does not have integration tests. Although not specifically mentioned in the requirements, we believe that having at least some integration tests are mandatory for an application of this complexity, because they check that all of the application components can work together.

No:
- The code does not have integration tests (bad structuring)
```

Reply:
I don't think example with integration testing is good even though in general I agree with the point. Problem with the example is that we do not ask for it therefore it should not be part of evaluation. We ask for tests and coverage, for example. We do not ask for _specific type of tests_. However, you do can argue that integration tests are needed (or most convenient way) to cover some very important paths. So you would end up with "Did not cover X. One way to do it would be using integration tests" or something like that.

I believe requirements are made open in order to give candidates freedom in terms of expressing their decision-making and coding culture rather than "we don't know what we want" or "we don't wanna say what we expect from you".